rspec-tools/rspec_tools/validation/description.py

```diff
@@ -1,55 +1,160 @@
+  sections = {}
+  optional_sections = {}
+  current_map = sections
+  current_section_name = ''
+  for line in education_format_file:
+    if line.startswith('== '):
+      section = line.replace('== ', '').strip()
+      if section.endswith('(optional)'):
+        section = section.replace(' (optional)', '')
+        current_map = optional_sections
+      else:
+        current_map = sections
+      if HOW_TO_FIX_IT_REGEX.match(section):
+        # we store "How to fix it" and "How to fix it in ..." together
+        section = HOW_TO_FIX_IT
+      current_section_name = section
+      if section not in current_map:
+        # If a section appears multiple times in the doc, we don't overwrite its subsections
+        current_map[section] = set()
+    if line.startswith('=== '):
+      section = line.replace('=== ', '').strip()
+      current_map[current_section_name].add(section)
+  return [
+    sections,
+    optional_sections,
+  ]
+
 # The list of all the sections currently accepted by the script.
 # The list includes multiple variants for each title because they all occur
 # in the migrated RSPECs.
 # Further work required to shorten the list by renaming the sections in some RSPECS
 # to keep only on version for each title.
-SECTION_NAMES_PATH = Path(__file__).parent.parent.parent.parent.joinpath('docs/section_names.adoc')
-SECTION_NAMES_FILE = SECTION_NAMES_PATH.read_text(encoding='utf-8').split('\n')
-ACCEPTED_SECTION_NAMES: Final[list[str]] = [s.replace('* ', '').strip() for s in SECTION_NAMES_FILE if s.strip()]
+LEGACY_SECTION_NAMES: Final[list[str]] = parse_names('docs/header_names/legacy_section_names.adoc')
 # The list of all the framework names currently accepted by the script.
-FRAMEWORK_NAMES_PATH = Path(__file__).parent.parent.parent.parent.joinpath('docs/allowed_framework_names.adoc')
-FRAMEWORK_NAMES_FILE = FRAMEWORK_NAMES_PATH.read_text(encoding='utf-8').split('\n')
-ACCEPTED_FRAMEWORK_NAMES: Final[list[str]] = [s.replace('* ', '').strip() for s in FRAMEWORK_NAMES_FILE if s.strip()]
+ACCEPTED_FRAMEWORK_NAMES: Final[list[str]] = parse_names('docs/header_names/allowed_framework_names.adoc')
+
+[
+  SECTIONS,
+  OPTIONAL_SECTIONS
+  ] = parse_education_section_names('docs/header_names/education_format_names.adoc')
+
+
+
+def intersection(lst1, lst2):
+  lst3 = [value for value in lst1 if value in lst2]
+  return lst3
+def difference(lst1, lst2):
+  return list(set(lst1) - set(lst2))
 
 def validate_section_names(rule_language: LanguageSpecificRule):
+  """Validates all h2-level section names"""
+
   descr = rule_language.description
-  for h2 in descr.find_all('h2'):
-    name = h2.text.strip()
-    if name not in ACCEPTED_SECTION_NAMES:
-      raise RuleValidationError(f'Rule {rule_language.id} has unconventional header "{name}"')
+  h2_titles = list(map(lambda x: x.text.strip(), descr.find_all('h2')))
+
+  education_titles = intersection(h2_titles, list(SECTIONS.keys()) + list(OPTIONAL_SECTIONS.keys()))
```

Comment:
If you change the `intersection` second parameter type to a set, you can write this shorter:
```suggestion
  education_titles = intersection(h2_titles, SECTIONS.keys() + OPTIONAL_SECTIONS.keys())
```

Reply:
I'm getting the following error:

```
>     education_titles = intersection(h2_titles, SECTIONS.keys() + OPTIONAL_SECTIONS.keys())
E     TypeError: unsupported operand type(s) for +: 'dict_keys' and 'dict_keys'
```

Reply:
Nevermind then, sorry. I thought `dict.keys()` returns a set, but it returns `dict_keys`